### ISSUE TITLE
Code generated by scaffolder leads to: Unable to cast object of type 'System.Boolean' to type 'System.SByte'

### ISSUE THREAD

**Issue by MKI-Miro** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#866]
Hi

I use dotnet core MySql package 3.0.0-rc1.final

In DB I have columns as tinyint(1) and they are scaffolded as sbyte, but when I make select i got this error: Unable to cast object of type 'System.Boolean' to type 'System.SByte'.'

Is it possible to fix it? or do I something wrong?

Thanks

**Comment by lauxjpn** [comment]
> Just tried Pomelo provider version 3.0.0-rc1.final to scaffold my simple MySQL 8.0.17 database with 9 tables using .NET Core 3.0, VS2019 16.3.3 and it worked without any issues. I used the same Scaffold-DbContext command from the above comment.

do you have tinyint(1) bolumn? (in phpmyadmin it is alias for boolean)

_Originally posted by @MKI-Miro in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/867#issuecomment-540648279_

**Comment by lauxjpn** [comment]
I have several columns that I create with:
TINYINT UNSIGNED NOT NULL DEFAULT 0
but I see that using MySQL Workbench, it shows them as:
tinyint(3) unsigned NOT NULL DEFAULT '0'

_Originally posted by @baronics in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/867#issuecomment-540650426_

**Comment by lauxjpn** [comment]
I can confirm that tiny int(4) works....  but tiny int(1) not

_Originally posted by @MKI-Miro in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/867#issuecomment-540661648_

**Comment by lauxjpn** [comment]
@MKI-Miro See the typemapper which implemented in Pomelo:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs

tinyint is mapping to sbyte, not bool.

_Originally posted by @yukozh in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/867#issuecomment-540665022_

**Comment by lauxjpn** [comment]
@yukozh yes correct but I think you didnt read carefully...

In db I have tinyInt(1) ->  this is scaffolded to sbyte (dotnet type) -> on select query it throws error

When I change it tinyInt(4) -> it scaffold also to sybte -> select query works

Difference is in context file 
.HasColumnType("tinyint(1)");
.HasColumnType("tinyint(4)");

_Originally posted by @MKI-Miro in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/867#issuecomment-540670826_

**Comment by bgrainger** [comment]
In case it's relevant: The underlying DB library, MySqlConnector, by default treats `TINYINT(1)` (which in MySQL is an alias for `BOOL`) as `System.Boolean`. Any other `TINYINT(n)` column is `System.SByte`.

**Comment by MKI-Miro** [comment]
> In case it's relevant: The underlying DB library, MySqlConnector, by default treats `TINYINT(1)` (which in MySQL is an alias for `BOOL`) as `System.Boolean`. Any other `TINYINT(n)` column is `System.SByte`.

So pomelo provider should be fixed? I am right?

**Comment by lauxjpn** [comment]
We can't really say yet. I will try to reproduce this issue in a couple of hours, if nobody else has picked it up until then.

**Comment by bhavukbhardwaj** [comment]
Hi, I am also having the same issue.

An unhandled exception occurred while processing the request.
   InvalidCastException: Unable to cast object of type &#x27;System.Boolean&#x27; to type &#x27;System.SByte&#x27 MySql.Data.MySqlClient.MySqlDataReader.GetFieldValue&lt;

**Comment by MKI-Miro** [comment]
@lauxjpn have you had time to fix it?

**Comment by lauxjpn** [comment]
I'll check it out after polishing the PR for #864.

**Comment by lauxjpn** [comment]
This seems to be a bug in the Pomelo provider.

Let's assume the following database table:

```sql
CREATE TABLE `ScaffoldMe` (
  `ScaffoldMeId` INT NOT NULL AUTO_INCREMENT,
  `NonNullableTinyInt1` TINYINT(1) NOT NULL,
  PRIMARY KEY (`ScaffoldMeId`)
);

INSERT INTO `ScaffoldMe` (`NonNullableTinyInt1`) VALUES ('1');
```

> https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/EFCore.MySql/Storage/Internal/MySqlTypeMappingSource.cs
>
> tinyint is mapping to sbyte, not bool.

As @yukozh points out, that is the one of the causes here. We map the `System.Boolean` to `bit(1)` and `System.SByte` to `tinyint` by default.

The following code, that was generated by the scaffolder, does not work with `TreatTinyAsBoolean=true` (the default):

```c#
public partial class Issue866Context : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        if (!optionsBuilder.IsConfigured)
        {
            optionsBuilder.UseMySql(
"server=127.0.0.1;user id=root;password=;port=3306;database=Issue866;TreatTinyAsBoolean=true");
        }
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<ScaffoldMe>(entity =>
        {
            entity.Property(e => e.ScaffoldMeId).HasColumnType("int(11)");
            entity.Property(e => e.NonNullableTinyInt1).HasColumnType("tinyint(1)");
        }
    }
}

public partial class ScaffoldMe
{
    public int ScaffoldMeId { get; set; }
    public sbyte NonNullableTinyInt1 { get; set; }
}

internal class Program
{
    private static void Main()
    {
        using var context = new Issue866Context();

        var result = context.Set<ScaffoldMe>().Single();

        Debug.Assert(result.NonNullableTinyInt1 == 1);
    }
}
```
It does not work, because MySqlConnector treats `tinyint(1)` as `System.Boolean`, but the scaffolder has defined `ScaffoldMe.NonNullableTinyInt1` as `System.SByte` and the MySqlConnect method `MySqlDataReader.GetFieldValue<T>(int ordinal)` is being used, that does not explicitly convert between primitive types.

**Now it get's tricky:**

> When I change it tinyInt(4) -> it scaffold also to sybte -> select query works
>
>Difference is in context file
>.HasColumnType("tinyint(1)");
>.HasColumnType("tinyint(4)");

As @MKI-Miro states correctly, overriding the specified type mapping by the scaffolder, is a viable option to force Pomelo to assume a specific type.

If you *don't* change the fluent model definition, but just `NonNullableTinyInt1` to `bool` in the model class, you still get the same error:

```c#
public partial class ScaffoldMe
{
    public int ScaffoldMeId { get; set; }
    public bool NonNullableTinyInt1 { get; set; } // <-- changed from sbyte to bool
}
```

That is, because MySqlConnector still reads `tinyint(1)` as a `System.Boolean` and Pomelo still expects a `System.SByte` as a result, because the fluent model definition forces Pomelo to assume a `tinyint(1)`, which is hard-mapped to `System.SByte`.

If you *do* change the model class and fluent model definition like the following, it *does* work:

```c#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<ScaffoldMe>(entity =>
    {
        entity.Property(e => e.ScaffoldMeId).HasColumnType("int(11)");
        entity.Property(e => e.NonNullableTinyInt1); // <-- removed HasColumnType
    }
}

public partial class ScaffoldMe
{
    public int ScaffoldMeId { get; set; }
    public bool NonNullableTinyInt1 { get; set; } // <-- changed from sbyte to bool
}
```

That is, because MySqlConnector still reads `tinyint(1)` as a `System.Boolean`, but Pomelo now falls back on it's default mapping for `System.Boolean` (which is `bit(1)`), because we did not set the column type explicitly. Though MySqlConnector and Pomelo expect different database column types, this still works, because both expect the same resulting CLR type.

Therefore, explicitly mapping the property to `bit(1)` (even though it is defined in the database as `tinyint(1)`) would lead to the same working outcome, because it is what Pomelo did under the hood in the previous case anyway:

```c#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<ScaffoldMe>(entity =>
    {
        entity.Property(e => e.ScaffoldMeId).HasColumnType("int(11)");
        entity.Property(e => e.NonNullableTinyInt1).HasColumnType("bit(1)"); // <-- bit(1)
    }
}

public partial class ScaffoldMe
{
    public int ScaffoldMeId { get; set; }
    public bool NonNullableTinyInt1 { get; set; } // <-- changed from sbyte to bool
}
```

**Now it gets trickier:**

When you set `TreatTinyAsBoolean=false` (the opposite of for what it is intended), suddenly the original scaffolded model just works.

It works for `public sbyte NonNullableTinyInt1 { get; set; }`, because now MySqlConnector and Pomelo both assume `System.SByte` as the return type.

It works for `public bool NonNullableTinyInt1 { get; set; }`, because the fluent model definition still explicitly sets the column type to `tinyint(1)`, which is hard-mapped to `System.SByte` and is therefore still the same type that MySqlConnector now assumes for `NonNullableTinyInt1`.

It surprisingly also works for `public bool NonNullableTinyInt1 { get; set; }`, if you ommit the explicit column type definition (or set it to `bit(1)`), because then, even though Pomelo now expects a `System.Boolean` and MySqlConnector reads a `System.SByte`, the `MySqlDataReader.GetBoolean(int ordinal)` method of MySqlConnector is called this time, which gracefully converts from `System.SByte` to `System.Boolean`.

**Quick workarounds:**

Either of the following workarounds will do for now:

- Change your model property to `bool` and remove the corresponding fluent model definition column type.
- Use `TreatTinyAsBoolean=false` and rely on type conversions by MySqlConnector.

Though the second option is quicker, the first one is more robust and should continue to work even after upcoming changes to Pomelo's type mapping model.

**Proposed changes:**

The actual scaffolding issue should be fixed in accordance with #811, which should fix the rest of the causes. It is important to consider the fact, that `TreatTinyAsBoolean=false` is widely used as a workaround at the moment.
The scaffolding process should respect the options specified by the user (e.g. the `TreatTinyAsBoolean` option), when reverse engeneering the model from the database.

**Comment by bhavukbhardwaj** [comment]
"Change your model property to bool and remove the corresponding fluent model definition column type "

This worked for me.. Thank You..!!

**Comment by lauxjpn** [comment]
@bgrainger Is the default CLR mapping for `bit` with `TreatTinyAsBool=true` in MySqlConnector `System.UInt64`?

**Comment by bgrainger** [comment]
Yes, all `BIT` columns are mapped to `ulong`.

**Comment by lauxjpn** [comment]
@MKI-Miro Thanks for taking the time to report this issue.
This has been fixed in #875.

**Comment by MKI-Miro** [comment]
@lauxjpn thank you for fixing
